### PR TITLE
fix pushAndRemoveUntil incorrectly removes the routes below the first…

### LINKED ISSUE #56688: pushAndRemoveUntil removing initialRoute in navigation stack after upgrading to v1.17
## Steps to Reproduce

1. Run `flutter create bug`.
2. Replace lib/main.dart with

```
import 'package:flutter/material.dart';

const homepage = '/';
const pageB = '/pageB';
const pageC = '/pageC';
const pageD = '/pageD';

final routes = <String, WidgetBuilder>{
  homepage: (BuildContext context) => Homepage(),
  pageB: (BuildContext context) => PageB(),
  pageC: (BuildContext context) => PageC(),
  pageD: (BuildContext context) => PageD(),
};

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      initialRoute: homepage,
      routes: routes,
    );
  }
}

class Homepage extends StatelessWidget {
  const Homepage({Key key}) : super(key: key);
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Homepage'),
      ),
      body: Center(
        child: FlatButton(
          child: Text('Go to Page B'),
          color: Colors.blue,
          onPressed: () {
            Navigator.pushNamed(context, pageB);
          },
        ),
      ),
    );
  }
}

class PageB extends StatelessWidget {
  const PageB({Key key}) : super(key: key);
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Page B'),
      ),
      body: Center(
        child: FlatButton(
          child: Text('Go to Page C'),
          color: Colors.blue,
          onPressed: () {
            Navigator.pushNamed(context, pageC);
          },
        ),
      ),
    );
  }
}

class PageC extends StatelessWidget {
  const PageC({Key key}) : super(key: key);
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Page C'),
      ),
      body: Center(
        child: FlatButton(
          child: Text('Go to Page D and remove until Page B'),
          color: Colors.blue,
          onPressed: () {
            Navigator.pushNamedAndRemoveUntil(
                context, pageD, ModalRoute.withName(pageB));
          },
        ),
      ),
    );
  }
}

class PageD extends StatelessWidget {
  const PageD({Key key}) : super(key: key);
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Page D'),
      ),
      body: Center(
        child: FlatButton(
          child: Text('Go back to Page B'),
          color: Colors.blue,
          onPressed: () {
            Navigator.pop(context);
          },
        ),
      ),
    );
  }
}
```

3. Run app and go until page D then go back to initial page

**Expected results:** 
I'm expecting being able to go back to the initialRoute (homepage) 

**Actual results:** 
I'm only able to go back till Page B as calling pushNamedAndRemoveUntil seems to also have removed the initialRoute (homepage)
<details>
  <summary>Logs</summary>

```
[✓] Flutter (Channel stable, v1.17.0, on Mac OS X 10.15.4 19E287, locale en-CN)
    • Flutter version 1.17.0 at /Users/christophewang/dev/flutter
    • Framework revision e6b34c2b5c (6 days ago), 2020-05-02 11:39:18 -0700
    • Engine revision 540786dd51
    • Dart version 2.8.1

 
[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    • Android SDK at /Users/christophewang/Library/Android/sdk
    • Platform android-29, build-tools 28.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b4-5784211)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.4.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.4.1, Build version 11E503a
    • CocoaPods version 1.9.1

[✓] Android Studio (version 3.6)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 44.0.2
    • Dart plugin version 192.7761
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b4-5784211)

[✓] VS Code (version 1.45.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.10.1

[✓] Connected device (1 available)
    • iPhone Xʀ • AD185E2B-3525-49C1-A434-72B9CB85A2A1 • ios • com.apple.CoreSimulator.SimRuntime.iOS-12-2 (simulator)

• No issues found!
```

</details>

### PR DESCRIPTION
**Pull request by chunhtai**
… matching route

## Description

The navigator refactor accidentally change the behavior of pushAndRemoveUntil method. Instead of stopping at the first route that match the predicate, it continue scanning through the entire route history and removing all the routes that does not match the predicate.

This pr fixes it.

## Related Issues

Fixes https://github.com/flutter/flutter/issues/56688

## Tests

I added the following tests:

See files

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I signed the [CLA].
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] All existing and new tests are passing.
- [ ] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [ ] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [ ] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: *Replace with a link to your migration guide*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/test/widgets/navigator_test.dart [hunk 1]
```diff
@@ -682,6 +682,42 @@ void main() {
     expect(find.text('B'), isOnstage);
   });
 
+  testWidgets('pushAndRemoveUntil does not remove routes below the first route that pass the predicate', (WidgetTester tester) async {
+    // Regression https://github.com/flutter/flutter/issues/56688
+    final GlobalKey<NavigatorState> navigator = GlobalKey<NavigatorState>();
+    final Map<String, WidgetBuilder> routes = <String, WidgetBuilder>{
+      '/' : (BuildContext context) => const Text('home'),
```
Comment (goderbauer) [inline packages/flutter/test/widgets/navigator_test.dart:689]:
```suggestion
      '/': (BuildContext context) => const Text('home'),
```

**Review by goderbauer — APPROVED** [review]
LGTM

**Comment by goderbauer** [comment]
Did this bug make it into the latest stable? We should consider a cherry-pick if it did since this could potentially break apps pretty badly...

**Comment by chunhtai** [comment]
> Did this bug make it into the latest stable? We should consider a cherry-pick if it did since this could potentially break apps pretty badly...

unfortunately yes... cc @tvolkert Is there a process to evaluate whether we should push a hot fix for this bug?

**Comment by tvolkert** [comment]
@chunhtai the process is to apply the cherry-pick request label to the issue.  I just labeled #56688 accordingly.  @pcsosinski FYI

**Comment by mxa0079** [comment]
When can we expect this to fix to make it to the stable channel? We are considering rolling back to 1.12 due to this breaking change.

**Comment by pcsosinski** [comment]
@mxa0079 this fix should be in 1.17.1+

**Pull request merged**